solidity/ecdsa/contracts/libraries/DKG.sol

```diff
@@ -0,0 +1,508 @@
+// SPDX-License-Identifier: MIT
+//
+// ▓▓▌ ▓▓ ▐▓▓ ▓▓▓▓▓▓▓▓▓▓▌▐▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓ ▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓ ▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▄
+// ▓▓▓▓▓▓▓▓▓▓ ▓▓▓▓▓▓▓▓▓▓▌▐▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓ ▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓ ▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓
+//   ▓▓▓▓▓▓    ▓▓▓▓▓▓▓▀    ▐▓▓▓▓▓▓    ▐▓▓▓▓▓   ▓▓▓▓▓▓     ▓▓▓▓▓   ▐▓▓▓▓▓▌   ▐▓▓▓▓▓▓
+//   ▓▓▓▓▓▓▄▄▓▓▓▓▓▓▓▀      ▐▓▓▓▓▓▓▄▄▄▄         ▓▓▓▓▓▓▄▄▄▄         ▐▓▓▓▓▓▌   ▐▓▓▓▓▓▓
+//   ▓▓▓▓▓▓▓▓▓▓▓▓▓▀        ▐▓▓▓▓▓▓▓▓▓▓         ▓▓▓▓▓▓▓▓▓▓         ▐▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓
+//   ▓▓▓▓▓▓▀▀▓▓▓▓▓▓▄       ▐▓▓▓▓▓▓▀▀▀▀         ▓▓▓▓▓▓▀▀▀▀         ▐▓▓▓▓▓▓▓▓▓▓▓▓▓▓▀
+//   ▓▓▓▓▓▓   ▀▓▓▓▓▓▓▄     ▐▓▓▓▓▓▓     ▓▓▓▓▓   ▓▓▓▓▓▓     ▓▓▓▓▓   ▐▓▓▓▓▓▌
+// ▓▓▓▓▓▓▓▓▓▓ █▓▓▓▓▓▓▓▓▓ ▐▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓ ▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓  ▓▓▓▓▓▓▓▓▓▓
+// ▓▓▓▓▓▓▓▓▓▓ ▓▓▓▓▓▓▓▓▓▓ ▐▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓ ▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓  ▓▓▓▓▓▓▓▓▓▓
+//
+//
+
+// Initial version copied from Keep Network Random Beacon:
+// https://github.com/keep-network/keep-core/blob/5138c7628868dbeed3ae2164f76fccc6c1fbb9e8/solidity/random-beacon/contracts/libraries/DKG.sol
+//
+// With a difference in group size which was set to 100.
+
+pragma solidity ^0.8.9;
+
+import "@openzeppelin/contracts/utils/cryptography/ECDSA.sol";
+import "@keep-network/sortition-pools/contracts/SortitionPool.sol";
+import "@keep-network/random-beacon/contracts/libraries/BytesLib.sol";
+import "../DKGValidator.sol";
+
+library DKG {
+    using BytesLib for bytes;
+    using ECDSA for bytes32;
+
+    struct Parameters {
+        // Time in blocks during which a submitted result can be challenged.
+        uint256 resultChallengePeriodLength;
+        // Time in blocks during which a result is expected to be submitted.
+        uint256 resultSubmissionPeriodLength;
```

Comment:
Yeah, great points! I wanted to work on it when removing `offchainDkgTime` as it will require some broader changes. Do you think it would be worth sorting it out completely in this PR?

Reply:
It depends on the approach we would like to take. If we want it to be close to the current implementation where we reset the timeout on a result challenge we should leave the name `resultSubmissionPeriodLength`.

If we change `resultSubmissionPeriodLength` to `dkgTimeout` it should be a very long time period that should also include possible challenges and we shouldn't be adjusting the submission block offset.

Reply:
We synced with @nkuba out of band and discussed the following steps are the best ones:

For ECDSA, we will introduce `resultSubmissionTimeout` governable parameter for DKG. This timeout will include the time for off-chain DKG and the time to submit the result for all group members (for instance, when the first 10 members are inactive). We will offset DKG timeout by the time at which the challenged result was submitted, minus DKG start time. We will maintain DKG result submission eligibility delay on-chain to make the off-chain client configuration easier, but this parameter will be available outside of DKG, probably alongside [other governable parameters defined in RFC 1](https://github.com/keep-network/tbtc-v2/blob/main/docs/rfc/rfc-1.adoc#3-governable-parameters).

For random beacon, we will not do any other modifications other than to remove eligibility checks for DKG result submission. Off-chain DKG time and eligibility checks will stay the same.

Reply:
fb14b3d6e3c4efab4db7d7eaef64666a9301349a